File: python algorithm/프로그래머스/LV2/게임 맵 최단거리.py
def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    queue = [(0, 0, 1)]
    maps[0][0] = 0
    
    while queue:
        x, y, dist = queue.pop(0)
        
        if x == n - 1 and y == m - 1:
            return dist

        for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
            nx, ny = x + dx, y + dy
            if 0 <= nx < n and 0 <= ny < m and maps[nx][ny] == 1:
                queue.append((nx, ny, dist + 1))
                maps[nx][ny] = 0        
            
    return -1

File: c++_algorithm/프로그래머스/Level2/bfs/게임맵 최단거리.cpp
#include <vector>
#include <queue>
using namespace std;

int dy[] = {-1,1,0,0};
int dx[] = {0,0,-1,1};
int bfs(vector<vector<int>> &maps, pair<int,int> start) {
    vector<vector<int>> visited = vector<vector<int>>(maps.size(), vector<int>(maps[0].size(), -1));
    queue<pair<int,int>> q;
    q.push(start);
    visited[start.first][start.second] = 1;
    
    while(!q.empty()) {
        int curr_row = q.front().first;
        int curr_col = q.front().second;
        q.pop();
        
        for (int dr = 0; dr < 4; dr++) {
            int ny = curr_row + dy[dr];
            int nx = curr_col + dx[dr];
            if (ny < 0 || nx < 0 || ny >= maps.size() || nx >= maps[0].size()) continue;
            if (visited[ny][nx] != -1) continue;
            if (maps[ny][nx] != 1) continue; 
            q.push({ny,nx});
            visited[ny][nx] = visited[curr_row][curr_col] + 1;
        }
    }
    return visited[maps.size() - 1][maps[0].size() - 1];
}

int solution(vector<vector<int> > maps)
{
    int answer = bfs(maps, {0,0});
    return answer;
}